src/internal/hooks/use-modal-component-analytics/index.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { useEffect } from 'react';
+
+import { PerformanceMetrics } from '../../analytics';
+import { ModalContextProps, useModalContext } from '../../context/modal-context';
+import { useEffectOnUpdate } from '../use-effect-on-update';
+
+export const usePrimaryButtonModalComponentAnalytics = (
+  isPrimaryButton: boolean,
+  elementRef: React.RefObject<HTMLElement>,
+  dependencies: React.DependencyList
+) => {
+  const modalContext = useModalContext();
+  useEffect(() => {
+    if (!isPrimaryButton || !elementRef.current || !modalContext.isInModal) {
+      return;
+    }
+    modalContext.componentLoadingCount.current++;
+    if (!isElementVisible(elementRef.current)) {
+      return;
+    }
+
+    modalContext.componentLoadingCount.current--;
+    emitModalContentReadyMetric(modalContext);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const isElementVisible = (htmlElement: HTMLElement) => {
+    if (!htmlElement) {
+      return true;
+    }
+    return (
+      htmlElement.offsetWidth > 0 &&
+      htmlElement.offsetHeight > 0 &&
+      getComputedStyle(htmlElement).visibility !== 'hidden'
+    );
+  };
+
+  useEffectOnUpdate(() => {
+    if (!isPrimaryButton || !elementRef.current || !modalContext.isInModal) {
+      return;
+    }
+    const elementVisible =
+      elementRef.current.offsetWidth > 0 &&
+      elementRef.current.offsetHeight > 0 &&
+      getComputedStyle(elementRef.current).visibility !== 'hidden';
+
+    if (!elementVisible) {
+      return;
+    }
+
+    modalContext.componentLoadingCount.current--;
```

Comment:
We're unconditionally decrementing the counter here, and I don't see that we're preventing this hook from decrementing the counter more than once per component instance. So whenever the dependencies of the hook change, the counter is decremented, which (if I'm not mistaken) can lead to a wrong value over time. It effectively would undo the counting of _other_ elements in the Modal

Reply:
Will update to add a function that relies on the loading state